### PR TITLE
<TBBAS-1802> Add support for exclusive control client

### PR DESCRIPTION
**Pull request by janm-dw**
- add support for exclusive control client
- add exclusive control drop others flag

### CODE REVIEW

**Review by NikolaiShipilov — CHANGES_REQUESTED** [review]
Lgtm overall, but it likely breaks the workaround put in place to prevent circular locks when destroying sessionHandler introduced for removing the temporary c-2-d streaming signals. Also added a few other minor comments.

File: shared/libraries/native_streaming_protocol/src/native_streaming_server_handler.cpp [hunk 1]
```diff
@@ -241,9 +243,16 @@ void NativeStreamingServerHandler::sendPacket(const SignalPtr& signal, PacketPtr
...
+    releaseSessionHandlerInternal(session, true);
+}
+
+void NativeStreamingServerHandler::releaseSessionHandlerInternal(SessionPtr session, bool enableSyncLock)
+{
+    std::shared_ptr<ServerSessionHandler> removedSessionHandler;  // keep object outside the scoped lock
     {
-        std::scoped_lock lock(sync);
+        std::unique_lock lock(sync, std::defer_lock);
+        if (enableSyncLock)
```
[the remark this answers is not in the document]
Comment (NikolaiShipilov) [reply shared/libraries/native_streaming_protocol/src/native_streaming_server_handler.cpp:254]: Yes, this solution should work fine

File: modules/tests/test_opendaq_device_modules/test_native_device_modules.cpp [hunk 2]
```diff
@@ -322,6 +323,125 @@ TEST_F(NativeDeviceModulesTest, ConnectUsernameDeviceAndStreamingConfig)
...
+    clientInstance = createAndConnectClient(ClientType::ExclusiveControl);
+    ASSERT_EQ(clientInstance.getDevices().getCount(), 1u);
+}
+
+TEST_F(NativeDeviceModulesTest, ClientTypeExclusiveControlDropOthers)
+{
+    auto serverInstance = InstanceBuilder().build();
+    serverInstance.addServer("OpenDAQNativeStreaming", nullptr);
+
+    auto createAndConnectClient = [](ClientType clientType)
```
[the remark this answers is not in the document]
Comment (janm-dw) [reply modules/tests/test_opendaq_device_modules/test_native_device_modules.cpp:415]: This slightly hides how to set ClientType on the instance, but I decided to do it anyway.

File: core/opendaq/opendaq/include/opendaq/client_type.h [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+#pragma once
+#include <coretypes/common.h>
+#include <coreobjects/property_object_ptr.h>
+
+BEGIN_NAMESPACE_OPENDAQ
+
+/*!
+ * @brief Enumeration of available client types
+ */
+enum class ClientType : EnumType
```
[the remark this answers is not in the document]
Comment (janm-dw) [reply core/opendaq/opendaq/include/opendaq/client_type.h:26]: I think the Exclusive control feature will also be supported on the OPC UA side in the future. So I would rather avoid adding the configuration prefix.

[the remark this answers is not in the document]
Comment (NikolaiShipilov) [reply core/opendaq/opendaq/include/opendaq/client_type.h:26]: Isn't OpcUa also considered a configuration protocol, or is 'configuration' intended specifically for nativeConfiguration ?

File: shared/libraries/native_streaming_protocol/src/native_streaming_server_handler.cpp [hunk 4]
```diff
@@ -454,6 +556,57 @@ void NativeStreamingServerHandler::onSessionError(const std::string& errorMessag
         });
 }
 
+void NativeStreamingServerHandler::releaseOtherControlConnectionsInternal(
+    std::shared_ptr<ServerSessionHandler> currentSessiohandler, std::vector<std::shared_ptr<ServerSessionHandler>> releasedSessionHanlders)
```
Comment (NikolaiShipilov) [inline shared/libraries/native_streaming_protocol/src/native_streaming_server_handler.cpp:560]: I guess the vector parameter should be passed by reference to make this work as intended. 

Could you also test this mechanism using `quick_start_simulator` and two instances of the GUI app?

Here are the steps:
* In the first GUI app, add a simulator device (with default values: "ClientType" - Control and "ExclusiveControlDropOthers" - false).
* Under the root of the first GUI app, add a local device `daqref://device0`.
* Under the simulated device, add a function block - a renderer is a good choice so you can view the connected signal.
* Connect one of the signals from the local device `daqref://device0` to this FB to trigger client-to-device streaming (you should see a display similar to the attached picture  c2d_renderer  ).
* Launch a second GUI app and add the simulated device with "ClientType" - ExclusiveControl and "ExclusiveControlDropOthers" - true. 
* The server should then drop the client connection made through the first GUI app, and the mirrored signal connected to the renderer should automatically be removed.

Comment (janm-dw) [reply shared/libraries/native_streaming_protocol/src/native_streaming_server_handler.cpp]: When connecting other exclusive control client with force drop flag, the streaming connection is not affected.
After some consideration, we have agreed that this is intended behaviour.

**Comment by euttsov-odaq** [comment]
1. If connect with exclusive client control from 2 clients, and both of them used "ExclusiveControlDropOthers" option, it does not handled properly in UI. Clients starting automatically reconnect and disconnect each other in cycle.

2. Empty error message in case if second client of type Control (0) or ExclusiveControl (1) type trying to connect to simulator

3. If connect to simulator via native protocol with ExclusiveControl client type, then you still be able to connect and control simulator via opcua from another client

**Comment by janm-dw** [comment]
1. Should be now fixed
2. Missing exception message is most likely caused by TBBAS-1888
3. OPC UA exclusive control will be implemented in a separate ticked

**Review by euttsov-odaq — APPROVED** [review]

**Pull request merged**